File: mobile/ios/ShiftPlanner/ShiftPlanner/Shared/Repositories/ScheduleRepository.swift
import Foundation

protocol ScheduleRepository {
    /// Generate a schedule for the period. Returns one entry per branch that was
    /// scheduled (the deployed backend returns a single object; algorithm2 an array).
    func generateSchedule(startDate: Date, endDate: Date, branchId: Int?) async throws -> [AppSchedule]

    func fetchSchedule(scheduleId: Int) async throws -> AppSchedule

    func fetchLatestSchedule(status: AppScheduleStatus?) async throws -> AppSchedule?

    /// Schedules (any status) overlapping the period, optionally for one branch.
    /// Used to detect conflicts before generating.
    func fetchSchedules(
        startDate: Date?,
        endDate: Date?,
        branchId: Int?,
        status: AppScheduleStatus?
    ) async throws -> [AppScheduleListItem]

    func publishSchedule(scheduleId: Int) async throws -> AppSchedule

    func fetchMySchedule() async throws -> [AppScheduledShift]

    func fetchAvailableEmployees(
        scheduleId: Int,
        shift: AppScheduledShift,
        branchId: Int?,
        includeUnavailable: Bool
    ) async throws -> [AppAvailableEmployee]

    func assignRequirement(
        scheduleId: Int,
        requirementId: Int,
        employeeId: Int
    ) async throws -> AppSchedule

    func createShift(scheduleId: Int, mutation: ScheduleShiftMutation) async throws -> AppSchedule

    func updateShift(
        scheduleId: Int,
        shiftId: Int,
        mutation: ScheduleShiftMutation
    ) async throws -> AppSchedule

    func deleteSchedule(scheduleId: Int) async throws

    func deleteShift(scheduleId: Int, shiftId: Int) async throws -> AppSchedule

    func updateScheduleRequirement(
        scheduleId: Int,
        requirementId: Int,
        mutation: ScheduleShiftMutation,
        quantity: Int
    ) async throws -> AppSchedule
}

File: mobile/android/app/src/main/java/com/froggyriia/shiftplanner/data/schedule/ScheduleRepository.kt
package com.froggyriia.shiftplanner.data.schedule

import com.froggyriia.shiftplanner.domain.model.AppAvailableEmployee
import com.froggyriia.shiftplanner.domain.model.AppSchedule
import com.froggyriia.shiftplanner.domain.model.AppScheduleListItem
import com.froggyriia.shiftplanner.domain.model.AppScheduledShift
import com.froggyriia.shiftplanner.domain.model.PendingShiftExchange
import com.froggyriia.shiftplanner.domain.model.ScheduleShiftMutation

interface ScheduleRepository {
    suspend fun generateSchedule(startDate: String, endDate: String, branchId: Int? = null): List<AppSchedule>
    suspend fun fetchSchedule(scheduleId: Int): AppSchedule
    suspend fun fetchLatestSchedule(status: String? = null): AppSchedule?
    /** Schedules (any status) that overlap the given period, optionally for one branch. */
    suspend fun fetchSchedules(
        startDate: String? = null,
        endDate: String? = null,
        branchId: Int? = null,
        status: String? = null
    ): List<AppScheduleListItem>
    suspend fun publishSchedule(scheduleId: Int): AppSchedule
    /**
     * The employee's own shifts. With a period given, shifts come from ALL
     * published schedules overlapping it (not just the latest one).
     */
    suspend fun fetchMySchedule(
        startDate: String? = null,
        endDate: String? = null
    ): List<AppScheduledShift>
    suspend fun fetchAvailableEmployees(
        scheduleId: Int,
        shift: AppScheduledShift,
        branchId: Int?,
        includeUnavailable: Boolean = false
    ): List<AppAvailableEmployee>
    suspend fun assignRequirement(
        scheduleId: Int,
        requirementId: Int,
        employeeId: Int
    ): AppSchedule
    suspend fun createShift(scheduleId: Int, mutation: ScheduleShiftMutation): AppSchedule
    suspend fun updateShift(
        scheduleId: Int,
        shiftId: Int,
        mutation: ScheduleShiftMutation
    ): AppSchedule
    suspend fun deleteSchedule(scheduleId: Int)
    suspend fun deleteShift(scheduleId: Int, shiftId: Int): AppSchedule
    suspend fun updateScheduleRequirement(
        scheduleId: Int,
        requirementId: Int,
        mutation: ScheduleShiftMutation,
        quantity: Int = 1
    ): AppSchedule

    // Shift exchange
    suspend fun createExchangeRequest(shiftId: Int, note: String)
    suspend fun fetchExchangeRequests(): List<PendingShiftExchange>
    suspend fun updateExchangeRequest(id: Int, approved: Boolean): PendingShiftExchange
}
